logictest: add newsession option to user directive

andyyang890 opened the pull request:
This patch adds a `newsession` option to the `user` directive,
which forces a new session to be created for the specified
`(username, nodeidx)` pair. This option is useful when we want
to verify a certain side effect has occurred after the current
session has ended and/or a fresh new session is created.

Informs #108751

Release note: None

andyyang890 committed:
logictest: add newsession option to user directive

This patch adds a `newsession` option to the `user` directive,
which forces a new session to be created for the specified
`(username, nodeidx)` pair. This option is useful when we want
to verify a certain side effect has occurred after the current
session has ended and/or a fresh new session is created.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/109109)
<!-- Reviewable:end -->

chrisseto commented on pkg/sql/logictest/logic.go:
```diff
@@ -375,12 +375,15 @@ import (
 //    Introduces a sleep period. Example: sleep 2s
 //
-//  - user <username> [nodeidx=N]
+//  - user <username> [nodeidx=N] [forcenew]
```
How would you feel about renaming this to `reconnect`, `re-establish`, `forcenewsession` or something of that nature? That seems a bit more inline with the documentation and provides some hinting about what the parameter does.

andyyang890 replied:
Renamed it to `newsession`!

chrisseto commented on pkg/sql/logictest/logic.go:
```diff
@@ -1170,19 +1173,27 @@ func (t *logicTest) outf(format string, args ...interface{}) {
+// if one exists. Otherwise, it opens and returns a new client. If forceNew is
+// specified, the existing client (if exists) will be closed and a new one will
+// be opened.
+func (t *logicTest) getOrOpenClient(user string, nodeIdx int, forceNew bool) *gosql.DB {
```
Do you think there'd be more future utility out of this functionality if you could instead of terminating and reopening a connection you exposed a session handle or name? Tests could then specify `user <name> session=notthedefault` to establish a new connection. Or is the important aspect terminating the active session rather than creating a new one?

andyyang890 replied:
Another PR I currently have open (that depends on this one) is to add a regression test for the temp view descriptor persistence bug we recently discovered: https://github.com/cockroachdb/cockroach/pull/109112. For that one, it is important that we end the previous session because we want to verify that the temp view's descriptor is dropped and this only happens after the session ends.

chrisseto reviewed (commented):
The change LGTM but the commit message would benefit from some additional motivation or example use cases. I understand what this commit is doing but not why it is doing it nor why it's opting for this particular approach.

andyyang890 commented:
@chrisseto Thanks for taking a look! I've updated the commit message and added a comment describing the use case for the option in the code. Also renamed the option from `forcenew` to `newsession`.

chrisseto reviewed (approved):
LGTM!
